test/connection.test.ts

```diff
@@ -68,46 +68,272 @@ describe('Connection', function () {
+    it('should complete sending all data from server when end is called on server side of the connection', async function() {
+      // TODO: Failing so skip for now
+      let data: Buffer[] = []
+      this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('data', (chunk: Buffer) => {
+          data.push(chunk)
+        })
+      })
+      const serverStream = this.serverConn.createStream()
+      serverStream.write(Buffer.alloc(30000))
+      await new Promise(setImmediate)
+      await this.serverConn.end()
+      assert.equal(Buffer.concat(data).length, 30000)
+    })
+
+    it('should complete sending all money from server when end is called on server side of the connection', async function () {
       const moneySpy = sinon.spy()
-      const dataSpy = sinon.spy()
+      let totalMoney: number = 0
       this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
-        stream.on('end', () => {
-          console.log('money stream end')
+        stream.setReceiveMax(12000)
+        stream.on('money', (amount) => {
           moneySpy()
+          totalMoney += +amount
         })
       })
-      this.clientConn.on('stream', (stream: DataAndMoneyStream) => {
-        stream.on('finish', () => {
-          console.log('data strem end')
-          dataSpy()
+      const serverStream = this.serverConn.createStream()
+      serverStream.setSendMax(6000)
+      await this.serverConn.end()
+      assert.equal(totalMoney, 12000)
+      assert.callCount(moneySpy, 6)
+    })
+
+    it('should complete sending all data from client when end is called on client side of the connection', async function() {
+      // TODO: Failing so skip for now
+      let data: Buffer[] = []
+      this.serverConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('data', (chunk: Buffer) => {
+          data.push(chunk)
+        })
+      })
+      const clientStream = this.clientConn.createStream()
+      clientStream.write(Buffer.alloc(30000))
+      await new Promise(setImmediate)
+      await this.clientConn.end()
+      assert.equal(Buffer.concat(data).length, 30000)
+    })
+
+    it('should complete sending all money from client when end is called on client side of the connection', async function () {
+      const moneySpy = sinon.spy()
+      let totalMoney: number = 0
+      this.serverConn.on('stream', (stream: DataAndMoneyStream) => {
+        stream.on('money', (amount) => {
+          moneySpy()
+          totalMoney += +amount
         })
       })
-      this.serverConn.createStream().setSendMax(100)
-      this.serverConn.createStream().write('hello')
+      const clientStream = this.clientConn.createStream()
+      clientStream.setSendMax(12000)
+      await this.clientConn.end()
+      assert.equal(totalMoney, 6000)
+      assert.callCount(moneySpy, 12)
+    })
+
+    it('should emit error on sending data from client after the server end is called', async function () {
```

Comment:
I think it would be better to just make the write call throw instead of emitting an error